source/dhtslib/memory.d

```diff
@@ -0,0 +1,77 @@
+module dhtslib.memory;
+
+import std.traits : isPointer, isSomeFunction, ReturnType;
+import core.lifetime : move;
+import std.typecons : RefCounted, RefCountedAutoInitialize;
+import htslib;
+
+struct HtslibMemory(T, alias destroy)
+if(!isPointer!T && isSomeFunction!destroy)
+{
+    /// Pointer Wrapper
+    struct HtslibPtr
+    {
+        /// data pointer
+        T * ptr;
+
+        /// no copying this as that could result
+        /// in premature destruction
+        @disable this(this);
+
+        /// destroy 
+        ~this()
+        {
+            /// if destroy function return is void 
+            /// just destroy
+            /// else if int
+            /// destroy then check return value 
+            /// else don't compile
+            static if(is(ReturnType!destroy == void))
+                destroy(this.ptr);
+            else static if(is(ReturnType!destroy == int))
+            {
+                auto success = destroy(this.ptr);
+                if(!success) hts_log_error(__FUNCTION__,"Couldn't destroy "~T.stringof~" data using function "~destroy.stringof);
+            }else{
+                static assert(0, "HtslibMemory doesn't recognize destroy function return type");
+            }
+        }
+    }
+
+    /// reference counted HtslibPtr
+    RefCounted!(HtslibPtr, RefCountedAutoInitialize.yes) rcPtr;
+
+    /// get underlying data pointer
+    @property nothrow @safe pure @nogc
+    ref inout(T*) truePtr() inout return
```

Comment:
I think you also want to make the `scope` because while `HtslibPtr` is not copyable,  `T*` trivially is, and you done want a reference gained from `truePtr` to outlive `this`. This might already be handled by the `return` attribute though, I'm not an expert in D's safety features.

Reply:
Looking into this, I am not completely sure how to achieve this either. I'm not sure either `scope` or `@live` are fleshed out enough to recognize when to keep `this` in scope if referencing a member's data. It would be nice if D had a `scope!this` I could attach to the returned reference. Either keeping `this` in scope or simply not compiling whenever I try to escape a reference to `this` with a longer lifetime than `this`.

Reply:
when in doubt, ask on the learn forum